pyggi/line/line.py

```diff
@@ -37,7 +23,9 @@ def __init__(self, target, ingredient):
         self.ingredient = ingredient
 
     def apply(self, program, new_contents, modification_points):
-        return program.do_replace(self, new_contents, modification_points)
+        engine = program.engines[self.target[0]]
+        assert engine == program.engines[self.ingredient[0]]
+        return engine.do_replace(program, self, new_contents, modification_points)
```

Comment:
Where is `do_replace()` defined?
It lacks an abstract class between the granularity-agnostic AbstractEngine and its implementation that would declare the methods `do_replace()` etc. This is even more visible for the tree-level granularity.

Reply:
Ok, then we should add one more class, for example, `AbstractTreeEngine`, between the `AbstractEngine` and `AstorEngine`/`SrcMLEngine`.

Reply:
Yes, we need "AbstractEngine > (Abstract)TreeEngine > AstorEngine" for the exact same reasons we had "AbstractProgram > (Abstract)TreeProgram > AstorProgram".

Reply:
Ok. Will fix it soon.

Reply:
commit: https://github.com/coinse/pyggi/pull/4/commits/8be595dd22973464e0426d7748aa395ae616057a